### PR TITLE
Exception customization support

### PR DESCRIPTION
**Pull request by jskeet**
(Prototype updated.)

### CODE REVIEW

**Comment by jskeet** [comment]
(Assume anything around naming is up for grabs, btw.)

File: src/Google.Api.Gax.Grpc/ApiCallExceptionExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,50 @@
+﻿/*
+ * Copyright 2016 Google Inc. All Rights Reserved.
```
Comment (chrisdunelm) [inline src/Google.Api.Gax.Grpc/ApiCallExceptionExtensions.cs]: You're still living in the past ;)

File: src/Google.Api.Gax.Grpc/ApiCallExceptionExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+                {
+                    return await fn(request, callSettings).ConfigureAwait(false);
+                }
+                catch (RpcException exception) when (customizer(exception, out var replacement))
+                {
+                    throw replacement;
+                }
+            };
+
+        // Sync retry
```
Comment (chrisdunelm) [inline src/Google.Api.Gax.Grpc/ApiCallExceptionExtensions.cs]: Wrong comment

**Comment by jskeet** [comment]
Fixed the two c'n'p errors.

**Comment by chrisdunelm** [comment]
Partial methods can't return values ;) , so the constructor sequence will be:

``` C#
_callBar = ...;
Modify_BarApiCall(ref _callBar);
```

and the partial method will be:

``` C#
partial void Modify_BarApiCall(ref ApiCall<BarRequest, BarResponse> call) =>
    call = call.WithExceptionCustomizer((RpcException original, out RpcException replaced) =>
    {
        if (original.Status.StatusCode == ...)
        {
            replaced = new MyException(original);
            return true;
        }
        replaced = null;
        return false;
    }
```

**Comment by jskeet** [comment]
Yup - that's what I meant :) Updated the PR comment for clarity.

**Comment by jskeet** [comment]
Updated as per conversation and added tests. PTAL.

File: src/Google.Api.Gax.Grpc/ApiCall.cs [hunk 3]
```diff
@@ -12,6 +12,15 @@
...
 namespace Google.Api.Gax.Grpc
 {
+    /// <summary>
+    /// Delegate representing exception customization. This allows specific RpcExceptions to be
+    /// customized, by returning a new RpcException to replace the original. If the original
+    /// exception should be passed through, the delegate should return <c>null</c>.
+    /// </summary>
+    /// <param name="original">The original exception thrown by gRPC.</param>
+    /// <returns>A replacement exception, or <c>null</c> if no customization should be performed.</returns>
+    public delegate RpcException ExceptionCustomizer(RpcException original);
```
Comment (chrisdunelm) [inline src/Google.Api.Gax.Grpc/ApiCall.cs]: Is this now required? Why not use a Func instead?

File: test/Google.Api.Gax.Grpc.Tests/ApiCallTest.cs [hunk 4]
```diff
@@ -116,5 +117,59 @@ private void AssertMetadata(Metadata expected, Metadata actual)
...
+        {
+            var response = new SimpleResponse();
+            var original = new RpcException(Status.DefaultCancelled);
+            var call = FakeApiCall.Create<SimpleRequest, SimpleResponse>((req, options) => throw original);
+            RpcException callbackException = null;
+            call = call.WithExceptionCustomizer(x => { callbackException = x; return null; });
+            var request = new SimpleRequest();
+            try
+            {
+                var ignored = sync ? call.Sync(request, null) : await call.Async(request, null);
```
Comment (chrisdunelm) [inline test/Google.Api.Gax.Grpc.Tests/ApiCallTest.cs]: Use Assert.Throws to ensure the exception really is thrown?

File: test/Google.Api.Gax.Grpc.Tests/ApiCallTest.cs [hunk 5]
```diff
@@ -116,5 +117,59 @@ private void AssertMetadata(Metadata expected, Metadata actual)
...
+            var original = new RpcException(Status.DefaultCancelled);
+            var replacement = new RpcException(Status.DefaultCancelled, "This is the replacement");
+            var call = FakeApiCall.Create<SimpleRequest, SimpleResponse>((req, options) => throw original);
+            RpcException callbackException = null;
+            call = call.WithExceptionCustomizer(x => { callbackException = x; return replacement; });
+            var request = new SimpleRequest();
+            try
+            {
+                var ignored = sync ? call.Sync(request, null) : await call.Async(request, null);
+            }
```
Comment (chrisdunelm) [inline test/Google.Api.Gax.Grpc.Tests/ApiCallTest.cs]: And here

File: src/Google.Api.Gax.Grpc/ApiCall.cs (hunk 3 above)
Comment (jskeet) [reply src/Google.Api.Gax.Grpc/ApiCall.cs]: Ooh, of course. Yay.

File: test/Google.Api.Gax.Grpc.Tests/ApiCallTest.cs (hunk 4 above)
Comment (jskeet) [reply test/Google.Api.Gax.Grpc.Tests/ApiCallTest.cs]: Yup, will do. Good catch, no pun intended.

**Comment by jskeet** [comment]
Right, changes made - PTAL :)

File: test/Google.Api.Gax.Grpc.Tests/ApiCallTest.cs [hunk 6]
```diff
@@ -116,5 +117,51 @@ private void AssertMetadata(Metadata expected, Metadata actual)
...
+        [Theory, CombinatorialData]
+        public async Task WithCustomExceptions_UncustomizedException(bool sync)
+        {
+            var response = new SimpleResponse();
+            var original = new RpcException(Status.DefaultCancelled);
+            var call = FakeApiCall.Create<SimpleRequest, SimpleResponse>((req, options) => throw original);
+            RpcException callbackException = null;
+            call = call.WithExceptionCustomizer(x => { callbackException = x; return null; });
+            var request = new SimpleRequest();
+            // It's slightly awkward to call specify the lambda in ThrowsAsync.
```
Comment (chrisdunelm) [inline test/Google.Api.Gax.Grpc.Tests/ApiCallTest.cs]: Nit: Slight odd grammar in this comment...

Comment (jskeet) [reply test/Google.Api.Gax.Grpc.Tests/ApiCallTest.cs]: Have modified it somewhat. Will merge with the new version, and we can fiddle more later.

**Pull request merged**